existential-vs-universal-quantification/article.md

```diff
@@ -0,0 +1,309 @@
+
+```haskell
+{-# LANGUAGE RankNTypes #-}
+
+func :: (forall a. Show a => a -> IO ()) -> IO ()
+```
+Now the compiler knows, that whatever function we pass to `func`, type variable `a` should be instantiated here, not outside.
+With this change we can instantiate `a` when we call log.
+
+```haskell
+func :: (forall a. Show a => a -> IO ()) -> IO ()
+func log = do
+  username <- getUsername
+  log @String username
+
+  userCount <- getUserCount
+  log @Int userCount
+```
+
+Note, that this type applications is unnecessary.
+Compiler is smart enough to choose appropriate types.
+
+#### Heterogeneous lists
+
+Now we understand that existential types allow you to introduce type variables without the necessity to specify them.
+Let's take a look how this is done in datatypes.
+You can introduce existentially quantified type variable in the datatype, by using `forall` on the left side of the constructor name. Existential types in datatypes allow you to have field in the datatype, without necessity to specify its type.
+
+```haskell
+data Elem = forall a. Elem a
+
+multitypedList :: [Elem]
+multitypedList = [Elem "a", Elem 1, Elem (Just 5)]
+```
+
+This is a heterogeneous list that contains values of different types.
+You can't do that with an universally quantified list because every element in that list would need to have the same type.
+But, there are not many things you can do with elements of this list.
+Since the type of elements is unknown at the use-site, you can't address it, nor can you pass it to functions that expect values of concrete type.
+`Data.Typeable` module's functionality can help to circumvent this restriction. It allows you to cast one type (even unknown) to another one.
+
+Adding constraints to the `a` variable may also improve the situation:
+
+```haskell
+-- You need to enable either `ExistentialQuantification` or `GADTs` extension to use this syntax
+data Elem = forall a. (Show a) => Elem a
+
+multitypedList :: [Elem]
+multitypedList = [Elem "a", Elem 1, Elem (Just 5)]
+
+-- We can use 'print' here because inner types of 'Elem' has 'Show' constraint
+printElem :: Elem -> IO ()
+printElem (Elem x) = print x
+
+main :: IO ()
+main = fmap printElem multitypedList
+
+λ> "a"
+λ> 1
+λ> Just 5
+```
+Note that since type variable `a` is hidden with a quantifier, you can't add constraints to it besides those added to the constructor.
+
+```haskell
+showFunc :: Show ??? => [Elem] -> String -- There is no type variable, that you can add constraint to
+showFunc (x:xs) => (show x) ++ (showFunc xs)
+```
+
+Other useful example of hiding type variables is `SomeException` wrapper
+
+```haskell
+data SomeException = forall a. Exception a => SomeException a
+```
+
+It allows you to catch exceptions of any type.
+As long, as you hadn't specified its type, (and it is not obvious to the compiler, from the handler for example), all exceptions will be caught, wrapped in `SomeException` datatype.
```

Comment:
I don't understand this sentence, there is something wrong with the way it is written.